### PR TITLE
ENG-688: sessions pagination

### PR DESCRIPTION
**Pull request by efiShtain**
Not done yet
please review the basic idea

This is stacked PR on ENG-702

### CODE REVIEW

File: sdk/sdkservices/sessions.go [hunk 1]
```diff
@@ -13,12 +13,22 @@ type ListSessionsFilter struct {
...
 	CountOnly    bool
+
+	PageSize  int32
+	Skip      int32
+	PageToken string
+}
+
+type ListSessionResult struct {
+	Sessions      []sdktypes.Session
+	Count         int
```
Comment (itayd) [inline sdk/sdkservices/sessions.go:24]: do we really need this? is there a way where we can ask only for the count?

Comment (itayd) [reply sdk/sdkservices/sessions.go:24]: oh there is. right.

File: sdk/sdkservices/sessions.go [hunk 2]
```diff
@@ -13,12 +13,22 @@ type ListSessionsFilter struct {
 	BuildID      sdktypes.BuildID
 	StateType    sdktypes.SessionStateType
 	CountOnly    bool
+
+	PageSize  int32
+	Skip      int32
+	PageToken string
```
Comment (itayd) [inline sdk/sdkservices/sessions.go]: i think these should be common to all paginated results. put these in a common struct and just reuse it all over.

File: sdk/sdkservices/sessions.go [hunk 3]
```diff
@@ -13,12 +13,22 @@ type ListSessionsFilter struct {
...
+	Sessions      []sdktypes.Session
+	Count         int
+	NextPageToken string
 }
 
 type Sessions interface {
 	Start(ctx context.Context, session sdktypes.Session) (sdktypes.SessionID, error)
 	Stop(ctx context.Context, sessionID sdktypes.SessionID, reason string, force bool) error
-	List(ctx context.Context, filter ListSessionsFilter) ([]sdktypes.Session, int, error)
+	List(ctx context.Context, filter ListSessionsFilter) (ListSessionResult, error)
```
Comment (itayd) [inline sdk/sdkservices/sessions.go:28]:
```suggestion
	List(ctx context.Context, filter ListSessionsFilter) ([]sdktypes.Session, PaginationInfo, error)
```

then you could reuse this everywhere.

**Comment by itayd** [comment]
didn't we talk also about secondary sorting?

File: sdk/sdkservices/sessions.go (hunk 3 above)
Comment (efiShtain) [reply sdk/sdkservices/sessions.go:31]: WDYM ? why can't I reuse it everywhere anyway?

File: internal/backend/db/dbgorm/sessions.go [hunk 4]
```diff
@@ -150,21 +150,43 @@ func (db *gormdb) listSessions(ctx context.Context, f sdkservices.ListSessionsFi
...
+	if f.Skip != 0 {
+		q = q.Offset(int(f.Skip))
+	}
+
+	if f.PageToken != "" {
+		q = q.Where("session_id < ?", f.PageToken)
+	}
+
+	// Double order in case we have two rows with the same created_at, then solve order by session_id
+	if err := q.
```
Comment (efiShtain) [inline internal/backend/db/dbgorm/sessions.go:166]: @itayd 
This is the double sorting

File: sdk/sdkservices/sessions.go (hunk 2 above)
Comment (itayd) [reply sdk/sdkservices/sessions.go:19]: add `TotalCount`

File: sdk/sdkservices/sessions.go (hunk 1 above)
Comment (itayd) [inline sdk/sdkservices/sessions.go]: `TotalCount`

Comment (efiShtain) [reply sdk/sdkservices/sessions.go]: Changed

File: sdk/sdkservices/sessions.go (hunk 3 above)
Comment (efiShtain) [reply sdk/sdkservices/sessions.go:28]: I did some changes, 
created an embeddable struct 
PTAL and let me know if that makes sense for you

in any case, let's ack this PR and open another ticket to change it if we want a different design

Comment (itayd) [reply sdk/sdkservices/sessions.go:28]: i still think these should be separate return vals, no need for the results struct.

I'll approve, but please fix that.

**Review by itayd — APPROVED** [review]

**Pull request merged**